pubsublite/streams.go

```diff
@@ -0,0 +1,333 @@
+	streamReconnecting  streamStatus = 1
+	streamConnected     streamStatus = 2
+	streamTerminated    streamStatus = 3
+)
+
+// streamHandler provides hooks for different streaming RPCs (e.g. publish,
+// subscribe, streaming cursor, etc) to use retryableStream. All Pub/Sub Lite
+// streaming RPCs implement a similar handshaking protocol, where an initial
+// request and response must be transmitted before other requests can be sent
+// over the stream.
+//
+// streamHandler methods must not be called while holding the retryableStream.mu
+// in order to prevent the streamHandler calling back into the retryableStream
+// and deadlocking.
+type streamHandler interface {
+	// newStream implementations must create the client stream with the given
+	// (cancellable) context.
+	newStream(context.Context) (grpc.ClientStream, error)
+	initialRequest() interface{}
+	validateInitialResponse(interface{}) error
+
+	// onStreamStatusChange is used to notify stream handlers when the stream has
+	// changed state. In particular, the `streamTerminated` state must be handled.
+	// retryableStream.Error() returns the error that caused the stream to
+	// terminate. Stream handlers should perform any necessary reset of state upon
+	// `streamConnected`.
+	onStreamStatusChange(streamStatus)
+	// onResponse forwards a response received on the stream to the stream
+	// handler.
+	onResponse(interface{})
+}
+
+// retryableStream is a wrapper around a bidirectional gRPC client stream to
+// handle automatic reconnection when the stream breaks.
+//
+// A retryableStream cycles between the following goroutines:
+//   Start() --> reconnect() <--> listen()
+// terminate() can be called at any time, either by the client to force stream
+// closure, or as a result of an unretryable error.
+type retryableStream struct {
+	// Immutable after creation.
+	ctx          context.Context
+	handler      streamHandler
+	responseType reflect.Type
+	timeout      time.Duration
+
+	// Guards access to fields below.
+	mu sync.Mutex
+
+	stream       grpc.ClientStream
+	cancelStream context.CancelFunc
+	status       streamStatus
+	finalErr     error
+}
+
+// newRetryableStream creates a new retryable stream wrapper. `timeout` is the
+// maximum duration for reconnection. `responseType` is the type of the response
+// proto received on the stream.
+func newRetryableStream(ctx context.Context, handler streamHandler, timeout time.Duration, responseType reflect.Type) *retryableStream {
+	return &retryableStream{
+		ctx:          ctx,
+		handler:      handler,
+		responseType: responseType,
+		timeout:      timeout,
+	}
+}
+
+// Start establishes a stream connection and returns the result to the given
+// channel.
+func (rs *retryableStream) Start(result chan error) {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+
+	if rs.status != streamUninitialized {
+		result <- status.Error(codes.Internal, "pubsublite: stream has already been started")
+		return
+	}
+	go rs.reconnect(&result)
```

Comment:
Channels don't need to be passed as a pointer as they are already a "reference" type like slices and maps.

Reply:
Looks like this method will also try to acquire the mutex.

Reply:
Oh yes, I forgot about channels being reference types, thanks! Fixed.

Yeah, rs.reconnect() does acquire the mutex, but I called it in a goroutine and was counting on it guaranteed to be asynchronous. Is there any chance it would run synchronously?

Reply:
I don't think that is guaranteed, but I could be wrong. @tbpg do you know?

Reply:
Shouldn't be a mutex issue. Even if the goroutine does block trying to acquire the mutex, the scheduler should let this function return (& release the mutex).

More fundamentally, I'm not sure about this API. What is the motivation for using this channel of errors approach?

Reply:
It will ultimately be used here:
https://github.com/tmdiep/google-cloud-go/blob/e9776305fceebc1ef1b637df2f1f3dc1be3f918b/pubsublite/publisher_internal.go#L506

We open multiple bidi rpc streams to publish pub/sub messages (one per topic partition). While starting up the publisher, I wanted to wait for all the streams to connect, but terminate all of them as soon as one returns an error, and also return the error.

Having said that, I can see another way to do this using callback funcs and will update this PR if it turns out simpler.

Reply:
I was able to implement the publisher startup logic without needing this result channel, so removed for simplification. PTAL, thanks!